// @elm-tooling/elm-language-server/out/compiler/utils/recordFieldReferenceTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecordFieldReferenceTable = void 0;
const syntaxNodeSet_1 = require("./syntaxNodeSet");
class RecordFieldReferenceTable {
    get frozen() {
        return this._frozen;
    }
    constructor(refsByField) {
        this._frozen = false;
        this.refsByField = refsByField !== null && refsByField !== void 0 ? refsByField : new Map();
    }
    static fromExpressions(fieldExpressions) {
        const fieldRefs = new Map();
        fieldExpressions.forEach((field) => {
            fieldRefs.set(field.name, new syntaxNodeSet_1.SyntaxNodeSet(field));
        });
        return new RecordFieldReferenceTable(fieldRefs);
    }
    get(field) {
        var _a, _b;
        return (_b = (_a = this.refsByField.get(field)) === null || _a === void 0 ? void 0 : _a.toArray()) !== null && _b !== void 0 ? _b : [];
    }
    addAll(other) {
        if (this._frozen || other.refsByField === this.refsByField) {
            return;
        }
        other.refsByField.forEach((refs, field) => {
            let set = this.refsByField.get(field);
            if (!set) {
                set = new syntaxNodeSet_1.SyntaxNodeSet();
                this.refsByField.set(field, set);
            }
            set.addAll(refs);
        });
    }
    plus(other) {
        const newRefs = new Map();
        this.refsByField.forEach((set, field) => {
            const newRefSet = new syntaxNodeSet_1.SyntaxNodeSet().addAll(set);
            const otherSet = other.refsByField.get(field);
            if (otherSet) {
                newRefSet.addAll(otherSet);
            }
            newRefs.set(field, newRefSet);
        });
        other.refsByField.forEach((set, field) => {
            if (!newRefs.get(field)) {
                newRefs.set(field, set);
            }
        });
        return new RecordFieldReferenceTable(newRefs);
    }
    isEmpty() {
        return this.refsByField.size === 0;
    }
    freeze() {
        this._frozen = true;
    }
}
exports.RecordFieldReferenceTable = RecordFieldReferenceTable;
