// @elm-tooling/elm-language-server/out/common/providers/diagnostics/diagnosticsRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiagnosticsRequest = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
class DiagnosticsRequest {
    static execute(requestor, files, onDone) {
        return new DiagnosticsRequest(requestor, files, onDone);
    }
    constructor(requestor, files, onDone) {
        this.files = files;
        this.done = false;
        this.token = new vscode_languageserver_1.CancellationTokenSource();
        if (!files.length) {
            this.done = true;
            onDone();
        }
        else {
            void requestor(files, 0, this.token.token).finally(() => {
                if (this.done) {
                    return;
                }
                this.done = true;
                onDone();
            });
        }
    }
    cancel() {
        if (!this.done) {
            this.token.cancel();
        }
        this.token.dispose();
    }
}
exports.DiagnosticsRequest = DiagnosticsRequest;
