// @elm-tooling/elm-language-server/out/common/providers/foldingProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FoldingRangeProvider = void 0;
const tsyringe_1 = require("tsyringe");
const vscode_languageserver_1 = require("vscode-languageserver");
const vscode_uri_1 = require("vscode-uri");
const elmWorkspaceMatcher_1 = require("../util/elmWorkspaceMatcher");
class FoldingRangeProvider {
    constructor() {
        this.REGION_CONSTRUCTS = new Set([
            "case_of_expr",
            "value_declaration",
            "type_alias_declaration",
            "type_declaration",
            "record_expr",
            "case_of_branch",
        ]);
        this.handleFoldingRange = (param) => {
            this.connection.console.info(`Folding ranges were requested`);
            const folds = [];
            const tree = param.sourceFile.tree;
            const findLastIdenticalNamedSibling = (node) => {
                // eslint-disable-next-line no-constant-condition
                while (true) {
                    const nextSibling = node.nextNamedSibling;
                    if ((nextSibling === null || nextSibling === void 0 ? void 0 : nextSibling.type) === "import_clause") {
                        node = nextSibling;
                    }
                    else {
                        return node;
                    }
                }
            };
            const treeCursor = tree.walk();
            const traverse = () => {
                var _a;
                const node = treeCursor.currentNode();
                if (((_a = node.parent) === null || _a === void 0 ? void 0 : _a.lastChild) && node.isNamed()) {
                    if ("import_clause" === node.type) {
                        const previousSibling = node.previousNamedSibling;
                        if (previousSibling === null ||
                            previousSibling.type !== "import_clause") {
                            const lastNode = findLastIdenticalNamedSibling(node);
                            folds.push({
                                endCharacter: lastNode.endPosition.column,
                                endLine: lastNode.endPosition.row,
                                kind: vscode_languageserver_1.FoldingRangeKind.Imports,
                                startCharacter: node.startPosition.column,
                                startLine: node.startPosition.row,
                            });
                        }
                    }
                    else if (node.type === "let_in_expr") {
                        // Use fields in the future
                        const valueDeclarations = node.namedChildren.filter((n) => n.type === "value_declaration");
                        const lastValueDeclaration = valueDeclarations[valueDeclarations.length - 1];
                        const letBody = node.lastNamedChild;
                        if (lastValueDeclaration) {
                            folds.push({
                                endCharacter: lastValueDeclaration.endPosition.column,
                                endLine: lastValueDeclaration.endPosition.row,
                                kind: vscode_languageserver_1.FoldingRangeKind.Region,
                                startCharacter: node.startPosition.column,
                                startLine: node.startPosition.row,
                            });
                        }
                        if (letBody) {
                            folds.push({
                                endCharacter: node.endPosition.column,
                                endLine: node.endPosition.row,
                                kind: vscode_languageserver_1.FoldingRangeKind.Region,
                                startCharacter: letBody.startPosition.column,
                                startLine: letBody.startPosition.row - 1,
                            });
                        }
                    }
                    else if (node.type === "if_else_expr") {
                        node.namedChildren.slice(1).forEach((child) => {
                            folds.push({
                                endCharacter: child.endPosition.column,
                                endLine: child.endPosition.row,
                                kind: vscode_languageserver_1.FoldingRangeKind.Region,
                                startCharacter: child.startPosition.column,
                                startLine: child.startPosition.row - 1,
                            });
                        });
                    }
                    else if (this.REGION_CONSTRUCTS.has(node.type)) {
                        folds.push({
                            endCharacter: node.endPosition.column,
                            endLine: node.endPosition.row,
                            kind: vscode_languageserver_1.FoldingRangeKind.Region,
                            startCharacter: node.startPosition.column,
                            startLine: node.startPosition.row,
                        });
                    }
                    else if ("block_comment" === node.type) {
                        folds.push({
                            endCharacter: node.endPosition.column,
                            endLine: node.endPosition.row,
                            kind: vscode_languageserver_1.FoldingRangeKind.Comment,
                            startCharacter: node.startPosition.column,
                            startLine: node.startPosition.row,
                        });
                    }
                }
                if (treeCursor.gotoFirstChild()) {
                    traverse();
                    while (treeCursor.gotoNextSibling()) {
                        traverse();
                    }
                    treeCursor.gotoParent();
                }
            };
            traverse();
            this.connection.console.info(`Returned ${folds.length} folding ranges`);
            return folds;
        };
        this.connection = tsyringe_1.container.resolve("Connection");
        this.connection.onFoldingRanges(new elmWorkspaceMatcher_1.ElmWorkspaceMatcher((param) => vscode_uri_1.URI.parse(param.textDocument.uri)).handle(this.handleFoldingRange.bind(this)));
    }
}
exports.FoldingRangeProvider = FoldingRangeProvider;
