// @elm-tooling/elm-language-server/out/common/util/sequence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sequence = void 0;
class Sequence {
    constructor(first, getter) {
        this.first = first;
        this.getter = getter;
        this.mappers = [];
        this.iterator = function* (first, getter, mappers) {
            const applyMappers = (value) => mappers.reduce((prev, mapper) => mapper(prev), value);
            yield applyMappers(first);
            let next = getter(first);
            while (next) {
                yield applyMappers(next);
                next = getter(next);
            }
        };
    }
    map(mapper) {
        const newSequence = new Sequence(this.first, this.getter);
        newSequence.mappers.push(mapper);
        return newSequence;
    }
    find(func) {
        const iterator = this.iterator(this.first, this.getter, this.mappers);
        let next = iterator.next();
        while (!next.done) {
            if (func(next.value)) {
                return next.value;
            }
            next = iterator.next();
        }
    }
    toArray() {
        const iterator = this.iterator(this.first, this.getter, this.mappers);
        const array = [];
        let next = iterator.next();
        while (!next.done) {
            array.push(next.value);
            next = iterator.next();
        }
        return array;
    }
}
exports.Sequence = Sequence;
