// @elm-tooling/elm-language-server/out/common/positionUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.positionEquals = exports.comparePosition = exports.PositionUtil = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
class PositionUtil {
    static FROM_VS_POSITION(position) {
        return new PositionUtil(position.line, position.character);
    }
    static FROM_TS_POSITION(position) {
        return new PositionUtil(position.row, position.column);
    }
    constructor(row, col) {
        this.row = row;
        this.col = col;
    }
    toVSPosition() {
        return vscode_languageserver_1.Position.create(this.row, this.col);
    }
    toTSPosition() {
        return {
            column: this.col,
            row: this.row,
        };
    }
}
exports.PositionUtil = PositionUtil;
function comparePosition(_pos1, _pos2) {
    // Convert TSPosition to VSPosition
    const pos1 = "row" in _pos1
        ? PositionUtil.FROM_TS_POSITION(_pos1).toVSPosition()
        : _pos1;
    const pos2 = "row" in _pos2
        ? PositionUtil.FROM_TS_POSITION(_pos2).toVSPosition()
        : _pos2;
    if (pos1.line === pos2.line && pos1.character === pos2.character) {
        return 0;
    }
    if (pos1.line < pos2.line ||
        (pos1.line === pos2.line && pos1.character < pos2.character)) {
        return -1;
    }
    return 1;
}
exports.comparePosition = comparePosition;
function positionEquals(a, b) {
    return a.line === b.line && a.character === b.character;
}
exports.positionEquals = positionEquals;
