// @elm-tooling/elm-language-server/out/common/protocol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProvideFileContentsRequest = exports.ReadDirectoryRequest = exports.ReadFileRequest = exports.FindTestsRequest = exports.GetDiagnosticsRequest = exports.UnexposeRequest = exports.ExposeRequest = exports.MoveRequest = exports.GetMoveDestinationRequest = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
// eslint-disable-next-line @typescript-eslint/no-namespace
exports.GetMoveDestinationRequest = new vscode_languageserver_1.RequestType("elm/getMoveDestinations");
exports.MoveRequest = new vscode_languageserver_1.RequestType("elm/move");
exports.ExposeRequest = new vscode_languageserver_1.RequestType("elm/expose");
exports.UnexposeRequest = new vscode_languageserver_1.RequestType("elm/unexpose");
exports.GetDiagnosticsRequest = new vscode_languageserver_1.RequestType("elm/getDiagnostics");
exports.FindTestsRequest = new vscode_languageserver_1.RequestType("elm/findTests");
exports.ReadFileRequest = new vscode_languageserver_1.RequestType("elm/readFile");
exports.ReadDirectoryRequest = new vscode_languageserver_1.RequestType("elm/readDirectory");
exports.ProvideFileContentsRequest = new vscode_languageserver_1.RequestType("elm/provideFileContents");
